### ISSUE TITLE
loadtxt/savetxt do not support radix 2

### ISSUE THREAD

**Issue by skoehler** [numpy/numpy#19836]
When processing large amount of data with Python, I rely on numpy to do the processing.
I seem to have no way of loading large text files with binary numbers in them, something like

```
101 11 111
110 01 000
```

I have found that `loadtxt` does handle the 0x prefix but not the 0b prefix (as available in Python), probably since you just use some strtoull function for the conversion from ASCII to numbers. On the other hand, the strtou* functions would allow me to specify a radix (2, 8, 10, 16, or even 13). This feature is currently not exposed.

I would like to request that I can force the radix. Also, it would be nice if numpy would handle the 0b to be more compatible with Python.

**Comment by rossbar** [comment]
You can provide custom conversion functions via the converters function:

```python
>>> import io
>>> fh = io.StringIO("101 11 111\n110 01 000")
>>> from functools import partial
>>> b2i = partial(int, base=2)
>>> np.loadtxt(fh, dtype=int, converters={i: b2i for i in range(3)})
array([[5, 3, 7],
       [6, 1, 0]])
```

The annoying part of this solution is that the `converters` argument has to be a dictionary that maps callables to column numbers. This means that you have to know the number of columns in the file a priori (hence the `range(3)` above). This could be made more convenient by adding #9291, but that's another issue.

**Comment by skoehler** [comment]
Yes, I am specifying converters right now. The more annoying part is the lack of performance. Calling a python function for each integer in the file being loaded somewhat defeats the purpose of numpy. Numpy is written in Python and should be fast not only at working with data but also at loading/saving data. Specifying python conversion functions will basically kill performance.

**Comment by rossbar** [comment]
> Numpy is written in Python and should be fast not only at working with data but also at loading/saving data. 

I think you meant "NumPy is written in C..." - which is true for many parts of the library, but [not loadtxt](https://github.com/numpy/numpy/blob/1fe57b2427e20f466c254ed6a98a014d9a9e574e/numpy/lib/npyio.py#L832). It's worth noting that the current implementation of `loadtxt` is *always* calling converters (i.e. Python functions) under-the-hood whether they're explicitly specified or not. In fact, the support for `0x`-prefixed strings that you reference in the OP is due to the fact that the [default converter](https://github.com/numpy/numpy/blob/1fe57b2427e20f466c254ed6a98a014d9a9e574e/numpy/lib/npyio.py#L724-L733) calls `float.fromhex` if float parsing initially fails. Supporting hex floats by default does result in a performance hit (see #19541), though this has recently been improved (see #19598).

All that is to say - Python functions are already being called on the string data during parsing in `loadtxt`, so in the current implementation there shouldn't be any major performance degradation from additional *call overhead* when using converters, e.g.

```python
In [1]: fh = ["101 11 111\n110 01 000"]

In [2]: %timeit np.loadtxt(fh)
11.1 µs ± 42.5 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [3]: converters = {i: b2i for i in range(3)}

In [4]: %timeit np.loadtxt(fh, dtype=int, encoding=None, converters=converters)  # encoding=None to skip an unnecessary internal encode/decode step
12.8 µs ± 50 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

At least some of the time difference is due to the fact that `float()` is faster than `int()` for these inputs.

**Comment by skoehler** [comment]
I apologice. I really thought that `loadtxt` was written in C. Given the fact that it is written in Python, using converters does not have a big impact. That much is true. But now I wish for a C implementation. After all, I can only process data efficiently with numpy if I can load/save it efficiently. Sorry, my previous experience with numpy was based on loading binary data.

I have also found, that `int("0b101", 0)` does work.

It's weird that you use `int(float(x))` as the default converter even if `dtype=np.uint64`. Doesn't Python use 64bit floats internally? So in conclusion converting to float first would actually subject the values in the files to rounding! Not a nice side-effect!
